models/event.js

```diff
@@ -22,6 +25,52 @@ function validateObject(object) {
   return cleanObject;
 }
 
+async function findAllRelatedEvents(id) {
+  const query = {
+    text: `
+    WITH searched_event AS (
+      SELECT
+        id,
+        metadata
+      FROM
+        events
+      WHERE
+        id = $1
+      LIMIT
+        1
+    ),
+    same_metadata_events AS (
+      SELECT
+        ARRAY_AGG(events.id) AS ids
+      FROM
+        events
+      INNER JOIN
+        searched_event ON true
+      WHERE
+        events.metadata = searched_event.metadata
```

Comment:
No comentário https://github.com/filipedeschamps/tabnews.com.br/pull/1638#discussion_r1607331459 do outro PR eu digo que os testes passariam se usasse `=` no lugar de `@>` para apontar que a implementação e os testes não estão considerando os casos reais que nem sempre serão simultâneos.

Até cito que talvez seja necessário usar algo como abaixo:

```sql
  events.metadata @> searched_event.metadata OR
  events.metadata <@ searched_event.metadata
```

Será que estou enxergando problema onde não existe?

Reply:
Isso, os testes consideram exatamente o caso que estou falando, mas os `expect`s estão apenas confirmando o resultado estranho, não o requisito que era desfazer ou confirmar todos os eventos relacionados passando o `id` de qualquer dos eventos.

Será que estamos interpretando de maneiras diferentes o que são "eventos relacionados" e a razão de ter sido criada a `findAllRelatedEvents`?

Reply:
> Será que estamos interpretando de maneiras diferentes o que são "eventos relacionados" e a razão de ter sido criada a `findAllRelatedEvents`?

A resposta para isso é fácil de descobrir: o array `events` retornado está sempre como você espera? Se sim, então estamos falando sobre as mesmas coisas e os eventos relacionados são esses mesmos.

Se os eventos retornados no `GET` não são os esperados, então te peço para me dizer um cenário de exemplo, assim eu consigo visualizar o mesmo problema que você, e então eu crio um teste que irá quebrar para poder corrigir esse comportamento.

> os `expect`s estão apenas confirmando o resultado estranho, não o requisito que era desfazer ou confirmar todos os eventos relacionados passando o `id` de qualquer dos eventos.

Se o único problema é na confirmação ou anulação dos eventos relacionados ao firewall, e o `GET` está correto, então o mais adequado é discutir isso no outro PR.

Reply:
Certo, vou criar o teste para o cenário `sucesso, sucesso, erro, sucesso, erro`, com a "passagem de tempo" para os conteúdos/usuários afetados não serem exatamente iguais.

Reply:
Para buscar dois eventos onde `metadata.contents` ou `metadata.users` são parcialmente contidos no evento que está sendo buscado, e que tenha a mesma `metadata.from_rule`, a comparação sugerida não parece funcionar. O SQL simplificado abaixo retorna apenas o evento buscado:

```sql
WITH searched_event AS (
  SELECT
    id,
    metadata
  FROM
    events
  WHERE
    id = $1
  LIMIT
    1
)
SELECT
  *
FROM
  events
INNER JOIN
  searched_event ON events.id = searched_event.id
WHERE
  events.metadata @> searched_event.metadata OR
  events.metadata <@ searched_event.metadata
``` 

O mais próximo que cheguei foi com uma comparação bem específica, do `metadata.contents` e `metadata.users`:

```sql
WITH searched_event AS (
  SELECT
    *
  FROM
    events
  WHERE
    id = $1
  LIMIT
    1
),
searched_event_users AS (
  SELECT
    jsonb_array_elements_text(metadata->'users') AS user_id
  FROM
    searched_event
),
searched_event_contents AS (
  SELECT
    jsonb_array_elements_text(metadata->'contents') AS content_id
  FROM
    searched_event
),
matching_events AS (
  SELECT
    events.*
  FROM
    events
LEFT JOIN searched_event ON true
  LEFT JOIN LATERAL (
    SELECT
      jsonb_array_elements_text(events.metadata->'users') AS user_id
  ) event_users ON TRUE
  LEFT JOIN LATERAL (
    SELECT
      jsonb_array_elements_text(events.metadata->'contents') AS content_id
  ) event_contents ON TRUE
  WHERE
  events.id <> searched_event.id AND (
    EXISTS (
      SELECT 1
      FROM searched_event_users
      WHERE searched_event_users.user_id = event_users.user_id
    )
    OR EXISTS (
      SELECT 1
      FROM searched_event_contents
      WHERE searched_event_contents.content_id = event_contents.content_id
    )
  )
)
SELECT * FROM searched_event
UNION ALL
SELECT * FROM matching_events
;
```

Não ficou uma query simples, e não é de fácil manutenção. Quando surgir mais um evento, precisará ser modificada adicionando um novo cursor, mais um `LEFT JOIN LATERAL` e mais um `OR EXISTS...`.

Não terminei as alterações no código e por isso não subi para o PR, mas decidi compartilhar a query para caso alguém saiba uma forma mais simples. Esse `SELECT` não verifica o `metadata.from_rule`, então seria mais uma condição para adicionar nele.

Reply:
> O SQL simplificado abaixo retorna apenas o evento buscado:

Na verdade, `@>` junto de `<@` busca apenas os eventos em que todos os afetados de um estão contidos no outro, e só um deles poderia ter itens a mais. Isso realmente não ia atender a todos os casos de teste que precisamos, pois também precisamos buscar casos de intersecção de afetados.

> Não terminei as alterações no código e por isso não subi para o PR, mas decidi compartilhar a query para caso alguém saiba uma forma mais simples.

Abaixo tem uma forma que acredito ser mais simples. E considerando os testes atuais, aparentemente só precisa ajustar um deles para não depender da ordem dos itens no array, pois a consulta não garante a ordem dos itens afetados. O teste `With two consecutive "firewall:block_users" events from the same IP`.

Mas como precisa criar os diversos casos de teste dos eventos relacionados, eu comecei a ver isso, mas percebi que é preciso modificar também a função `findByEventId` e as que ela chama, pois ali também não considera que os metadados de diferentes eventos podem não ser iguais, já que usa apenas os dados do último evento:

```js
const affectedData = await getAffectedData(relatedEvents.at(-1));
```

Não tive tempo de ajustar a `findByEventId`, mas deixo o exemplo de consulta que retorna todos os eventos de firewall relacionados:

```sql
    WITH RECURSIVE related_event AS (
      SELECT
        id,
        metadata->>'original_event_id' AS original_event_id,
        jsonb_array_elements_text(metadata->'contents') AS content_id,
        jsonb_array_elements_text(metadata->'users') AS user_id
      FROM
        events
      WHERE
        id = $1

    UNION

      SELECT
        e.id,
        e.metadata->>'original_event_id' AS original_event_id,
        jsonb_array_elements_text(e.metadata->'contents') AS content_id,
        jsonb_array_elements_text(e.metadata->'users') AS user_id
      FROM
        events e
      INNER JOIN related_event ON
        e.metadata->'contents' ? related_event.content_id OR
        e.metadata->'users' ? related_event.user_id
    )

    SELECT DISTINCT
      events.*
    FROM
      events
    INNER JOIN related_event ON
      events.id = related_event.id
```

Acredito que está fácil de adicionar outros campos de comparação, como `from_rule`, mas não parece ser necessário, e deve ser tranquilo lidar com isso na `findByEventId`.

Reply:
Obrigado, Felipe. Vou ler com atenção sua sugestão e corrijo o resto do código :+1:

Reply:
A query funciona e os testes asseguram isso. Realizei um teste inicial com 4 eventos no banco de dados e a consulta demorou 1 segundo; depois realizei mais alguns testes entre 1 mil e 25 mil eventos no banco de dados e o tempo não aumentou. Talvez precisemos adaptar algo para quando formos criar o endpoint que lista todos eventos (paginado), mas por ora acredito que está adequado :+1:

Reply:
Precisa criar índices para melhorar essa consulta, mas, por enquanto, por ser exclusivo da moderação, não vejo necessidade.

Sobre listar todos os eventos, daí acho que não precisa trazer os relacionados, então a consulta será mais simples.